src/ast/attachpoint_parser.cpp

```diff
@@ -0,0 +1,426 @@
+#include "attachpoint_parser.h"
+
+#include <algorithm>
+#include <exception>
+#include <functional>
+#include <string>
+#include <unordered_map>
+
+#include "types.h"
+
+namespace bpftrace {
+namespace ast {
+
+AttachPointParser::AttachPointParser(Program *root,
+                                     BPFtrace &bpftrace,
+                                     std::ostream &sink)
+    : root_(root), bpftrace_(bpftrace), sink_(sink)
+{
+}
+
+int AttachPointParser::parse()
+{
+  if (!root_)
+    return 1;
+
+  uint32_t failed = 0;
+  for (Probe *probe : *(root_->probes))
+  {
+    for (AttachPoint *ap : *(probe->attach_points))
+    {
+      if (parse_attachpoint(*ap))
+      {
+        ++failed;
+        bpftrace_.error(sink_, ap->loc, errs_.str());
+        errs_.str({}); // clear buffer
+      }
+    }
+  }
+
+  return failed;
+}
+
+int AttachPointParser::parse_attachpoint(AttachPoint &ap)
+{
+  ap_ = &ap;
+
+  parts_ = split_string(ap_->raw_input, ':', true);
+  if (parts_.empty())
+  {
+    errs_ << "Invalid attachpoint definition" << std::endl;
+    return 1;
+  }
+
+  ap_->provider = probetypeName(parts_.front());
+
+  switch (probetype(parts_.front()))
+  {
+    case ProbeType::kprobe:
+      return kprobe_parser();
+    case ProbeType::kretprobe:
+      return kretprobe_parser();
+    case ProbeType::uprobe:
+      return uprobe_parser();
+    case ProbeType::uretprobe:
+      return uretprobe_parser();
+    case ProbeType::usdt:
+      return usdt_parser();
+    case ProbeType::tracepoint:
+      return tracepoint_parser();
+    case ProbeType::profile:
+      return profile_parser();
+    case ProbeType::interval:
+      return interval_parser();
+    case ProbeType::software:
+      return software_parser();
+    case ProbeType::hardware:
+      return hardware_parser();
+    case ProbeType::watchpoint:
+      return watchpoint_parser();
+    default:
+      errs_ << "Unrecognized provider: " << ap_->provider << std::endl;
+      return 1;
+  }
+
+  return 0;
+}
+
+int AttachPointParser::kprobe_parser(bool allow_offset)
+{
+  if (parts_.size() != 2)
+  {
+    errs_ << ap_->provider << " provider requires 2 arguments" << std::endl;
+    return 1;
+  }
+
+  // Handle kprobe:func+0x100 case
+  auto plus_count = std::count(parts_[1].cbegin(), parts_[1].cend(), '+');
+  if (plus_count)
+  {
+    if (!allow_offset)
+    {
+      errs_ << "Offset not allowed" << std::endl;
+      return 1;
+    }
+
+    if (plus_count != 1)
+    {
+      errs_ << "Cannot take more than one offset" << std::endl;
+      return 1;
+    }
+
+    auto offset_parts = split_string(parts_[1], '+', true);
+    if (offset_parts.size() != 2)
+    {
+      errs_ << "Invalid offset" << std::endl;
+      return 1;
+    }
+
+    ap_->func = offset_parts[0];
+    try
+    {
+      ap_->func_offset = std::stoll(offset_parts[1], nullptr, 0);
```

Comment:
This (and other stoll/stoi) allows trailing non-numeric characters. e.g., `k:func+0x100zzz`.  It can be checked by

```cpp
std::size t idx;
ap_->func_offset = std::stoll(offset_parts[1], &idx, 0);
if (idx != offset_parts[1].size()) { /* err */ } 
```

When I implemented the support of attaching kprobe to offset, it took some time to understand attachpoint parser. This looks nice.